core/lib/zksync_core/src/lib.rs

```diff
@@ -335,6 +338,47 @@ pub async fn initialize_components(
         tokio::spawn(circuit_breaker_checker.run(stop_receiver.clone())),
     ];
 
+    if components.contains(&Component::DevConversionRateApi) {
+        let base_token_fetcher_config = configs
+            .base_token_fetcher
+            .clone()
+            .context("base_token_fetcher_config")?;
+
+        let stop_receiver = stop_receiver.clone();
+        let conversion_rate_task = tokio::spawn(async move {
+            dev_api_conversion_rate::run_server(stop_receiver, &base_token_fetcher_config).await
+        });
+        task_futures.push(conversion_rate_task);
+    };
+
+    let base_token_fetcher = if components.contains(&Component::BaseTokenFetcher) {
+        if components.contains(&Component::DevConversionRateApi) {
+            // If we are using the dev conversion rate API,
+            // we need to wait for it to start before initializing the base token fetcher.
+            // a more deterministic way should be used
+            tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
I don't think it's a very robust approach. Maybe you can provide something like:

```
impl BaseTokenFetcher {
// Will retry creation of the fetcher for at most `timeout`.
fn new_with_timeout(config: BaseTokenFetcherConfig, timeout: Duration) -> anyhow::Result<Self> {
 // ... try to create in a loop with small `sleep` until `timeout` passes or creation is successful
}
}
```

Still not pretty and shows fundamental problems in the design, but IMHO would suite as a better short-term approach.

Reply:
Done! Here's the implementation:
```rs
    /// Attemps to create a new `BaseTokenFetcher` instance with a timeout for getting the initial conversion rate
    pub async fn new_with_timeout(config: BaseTokenFetcherConfig) -> anyhow::Result<Self> {
        let http_client = reqwest::Client::new();

        let mut conversion_rate = None;
        let mut tries = 0;
        while conversion_rate.is_none() && tries < MAX_CONVERSION_RATE_FETCH_RETRIES {
            match http_client
                .get(format!(
                    "{}/conversion_rate/0x{}",
                    config.host,
                    config.token_address.encode_hex::<String>()
                ))
                .send()
                .await
            {
                Ok(res) => {
                    conversion_rate = Some(res.json::<u64>().await.context(
                        "Unable to parse the response of the native token conversion rate server",
                    )?);
                }
                Err(_err) => {
                    tries += 1;
                    sleep(Duration::from_secs(1)).await;
                }
            }
        }
```